central/sensor/enhancement/broker.go

```diff
@@ -0,0 +1,109 @@
+package enhancement
+
+import (
+	"context"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/stackrox/rox/central/sensor/service/connection"
+	"github.com/stackrox/rox/generated/internalapi/central"
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/concurrency"
+	"github.com/stackrox/rox/pkg/logging"
+	"github.com/stackrox/rox/pkg/sync"
+	"github.com/stackrox/rox/pkg/uuid"
+)
+
+var (
+	log = logging.LoggerForModule()
+)
+
+type enhancementSignal struct {
+	msgArrived concurrency.Signal
+	msg        *central.DeploymentEnhancementResponse
+}
+
+// The Broker coordinates and matches deployment enhancement requests to responses
+type Broker struct {
+	activeRequests map[string]*enhancementSignal
+	lock           sync.Mutex
+}
+
+// NewBroker returns a new broker
+func NewBroker() *Broker {
+	return &Broker{
+		activeRequests: make(map[string]*enhancementSignal),
+	}
+}
+
+// NotifyDeploymentReceived matches the ID of Sensors response to the request and calls the right callback for it
+func (b *Broker) NotifyDeploymentReceived(msg *central.DeploymentEnhancementResponse) {
+	if msg == nil || msg.GetMsg() == nil {
+		log.Warnf("Received empty message, skipping enhancement notify")
+		return
+	}
+
+	b.lock.Lock()
+	defer b.lock.Unlock()
+	reqID := msg.GetMsg().GetId()
+	sig, ok := b.activeRequests[reqID]
+	if !ok {
+		log.Warnf("Received response to an unknown Deployment Enrichment Request ID %s", reqID)
+		return
```

Comment:
Shouldn't we delete that from the map? Or is it okay to keep it and process it later?

Reply:
>  Maybe if central restarts while the enrichment request on Sensor is still running.

This is one of the case of `!ok`, and I think logging the warning and returning is the correct thing to do. 

From the current code, it's impossible for `b.activeRequests` to have the entry appended after the `ok` check, unless you make significant changes to `SendAndWaitForEnhancedDeployments`. 

I don't see any risks here and no need to assume the entry could be added in the meantime.